In [1]:
import random
import numpy as np
import matplotlib.pyplot as plt
import library_data_science as lds

# Inferential Statictics

A branch of statistics dealing with the problems of generalizing the results of a random sample study to the entire population and estimating the errors arising from such generalization.

![Illustration of Inferential Statistics](https://datatab.net/assets/tutorial/Descriptive_statistics_and_inferential_statistics.png)

**Population**: A set of examples.

**Sample**: A proper subset of a population.

# Monte Carlo simulation

A method of estimating the value of an unknown quantity using the principles of inferential statistics.

1. Set number of **tests**.

2. **Conduct** the set amount of tests.

3. The result is the **average value** of all tests.

So far, in virtually every notebook, I have come across **Monte Carlo** simulations.

* Probability that rolling five times we got combination `11111`.

* The birthday problem.

* Predicting the distance from the endpoint to the starting point using Random Walks.

# Law of Large Numbers

It states that if you repeat an experiment independently a large number of times and average the result, what you obtain should be close to expected value.

For example, we know that the mean of rolling a dice $n \geq 2$ times equals $3.5$. We can check the truth of the **Law of Large Numbers** using Monte Carlo simulation. I will repeat the test four times for bigger amount of trials and sum up the results.

In [2]:
dice = [1, 2, 3, 4, 5, 6]
trials = [10, 100, 1000, 10000000]

In [3]:
means = []

for num_trials in trials:
    results = []

    for attempt in range(num_trials):
        experiment = random.choice(dice)
        results.append(experiment)
    
    current_mean = lds.mean(results)
    means.append(current_mean)

print(means)

[3.5, 3.84, 3.424, 3.4998155]


|                 | Mean       |
|-----------------|------------|
| 10 Trials       | 3.5        |
| 100 Trials      | 3.84       |
| 1000 Trials     | 3.424      |
| 10000000 Trials | 3.4998155  |

# Calculus methods using Monte Carlo simulation

Instead, I will focus on a more practical application of the Monte Carlo simulation method. It is very useful for scientists who require computational methods for differential calculus in their calculations.

### Derivative

The first function is the method for calculating the derivative at a given point. According to the mathematical formula, the derivative can be computed using the equation below.

$$\frac{d}{dx}\bigg[ f(x_0) \bigg] = \lim_{\Delta x \to 0}{\frac{f(x_0 + \Delta x) - f(x_0)}{\Delta x}}$$

As we know, there are many more practical methods, but for computational purposes, it serves its role entirely. Writing a function to calculate the derivative at a point is not difficult.

In [4]:
def derivative(f, x_0: float, delta_x = 0.0001) -> float:
    """Function returns the derivative of the given function at specified point.
    
       Function must be continuous at specified point."""
    
    return (f(x_0 + delta_x) - f(x_0)) / delta_x

### Integral

The second function is the method for calculating the integral over a given interval. According to the mathematical formula, the integral can be computed using the riemann sum.

$$\int_{a}^{b}{f(x)}dx = \lim_{n \to +\infty}{\sum_{i=1}^{n}{f(c_i)\Delta x}}$$

Similarly, writing such a function in Python is not a significant challenge.

In [5]:
def integral(f, a: float, b: float, delta_x = 0.0001) -> float:
    """Function returns the integral of the given function at specified range.
    
       Function must be continuous at specified range."""
    
    counter = 0

    while a <= b:
        counter = counter + (f(a) * delta_x)
        a = a + delta_x
    
    return counter

However, this approach has many limitations. It assumes that the function is continuous over the specified interval. Accounting for discontinuities can be very challenging for beginner programmers and even for brilliant mathematicians. Of course, there are many libraries that account for such function behaviors.

### Area using Monte Carlo simulation

To estimate the area of some region $R$ follow the instruction.

1. Pick an enclosing region $E$, such that the area of $E$ is easy to calculate and $R$ lies completely within $E$.

2. Pick a set of random points that lie within $E$.

3. Let $F$ be the fraction of the points that fall within $R$. $$F = \frac{lying\_points}{all\_points}$$

4. Multiply the area of $E$ by $F$. That is the result. $$R \approx E \times F$$

In [6]:
def area_monte_carlo(f, a: float, b: float, c: float, d: float, num_samples = 1000000) -> float:
    """Function returns the integral of the given function at specified range.

       Method used in the process is Monte Carlo Simulation.

       Factors a, b are the range of x-axis, and factors c, d are the range of y-axis.
    
       Function must be continuous at specified range."""

    within_the_area = 0
    for sample in range(num_samples):
        x = random.uniform(a, b)
        y = random.uniform(c, d)

        try:
            if lds.distance((x, 0), (x, y)) <= f(x):
                within_the_area = within_the_area + 1
        except ZeroDivisionError:
            pass

    estimated_area = (np.abs(a - b) * np.abs(c - d)) * (within_the_area / num_samples)

    return estimated_area

Example for $\displaystyle\int_{0}^{\pi}{\sin{(x)}}dx$.

In [7]:
f = lambda x: np.sin(x)
num_samples = 100000
a = 0
b = np.pi
c = 0
d = 1

print('Estimated area of sin(x) on [0, pi]:', area_monte_carlo(f, a, b, c, d))
print('Real area of sin(x) on [0, pi]:', integral(f, a, b))

Estimated area of sin(x) on [0, pi]: 2.0004750956190263
Real area of sin(x) on [0, pi]: 1.9999999986725736


We can also easily calculate the value of $\pi$. To do this, we will compute the area of the upper part of a circle with a radius of 1. For this purpose, we use the function describing the upper half of the circle, $\sqrt{1 - x^2}$. This approach is called **Buffon-Laplace Method**.

In [8]:
f = lambda x: (1 - x ** 2) ** 0.5
num_samples = 100000
a = -1
b = 1
c = 0
d = 1

print('Estimated value of pi:', 2 * area_monte_carlo(f, a, b, c, d))
print('Real area of pi:', 2 * integral(f, a, b))

Estimated value of pi: 3.141912
Real area of pi: 3.141591477734107
